### PR TITLE
State management: Migrate to bloc pattern/library

### PR DESCRIPTION
**Pull request by bisgardo**
Migrate `ChangeNotifier` implementations to Cubits from the `flutter_bloc` library. Cubits were chosen over Blocs as the latter seems like it would just add the overhead of events (as compared to a direct function call) without added benefits that events might provide in other cases.

Compared to the implementaion based on Provider, this version doesn't "crash" when resetting the valid T&C version. Instead, a listener (on `BlocConsumer`) detects that the T&C acceptance state has changed and triggers a refresh. Being straightforward with Bloc, Proivder doesn't solve problems like this.

Unfortunately, we cannot get the listener to trigger on the initial value and therefore still need the component to be stateful (to use `initState`). The proper solution is probably to replace the initial loading of shared preferences with a more general initialization procedure which also fetches the valid T&C.

Apart from this difference, the two implementations are clearly very similar. Still, I find that bloc does add some niceties like `onChange` and more structure like a repository abstraction. And it of course have many features that we haven't even touched here.

In conclusion, Bloc seems to be strictly more powerful than Provider as the two solutions being very similar use basically the entire feature set of Provider but relatively little of that of Bloc.

### CODE REVIEW

**Comment by bisgardo** [comment]
Note that the PR is based on https://github.com/Concordium/concordium-wallet/pull/12 which has not yet been merged.

File: lib/screens/home/screen.dart [hunk 1]
```diff
@@ -19,25 +19,37 @@ class _HomeScreenState extends State<HomeScreen> {
...
   static Future<void> _updateValidTac(WalletProxyService walletProxy, TermsAndConditionAcceptance tacAcceptance) async {
     final tac = await walletProxy.fetchTermsAndConditions();
     tacAcceptance.validVersionUpdated(ValidTermsAndConditions.refreshedNow(termsAndConditions: tac));
   }
 
+  void _refresh(BuildContext context, TermsAndConditionAcceptance tacAcceptance) {
+    final network = context.read<ActiveNetwork>().state;
+    final services = context.read<ServiceRepository>().networkServices[network.active]!;
+    _updateValidTac(services.walletProxy, tacAcceptance);
+  }
```
Comment (schwartz-concordium) [inline lib/screens/home/screen.dart:36]:
```suggestion
  void _refresh(BuildContext context) {
    final network = context.read<ActiveNetwork>().state;
    final services = context.read<ServiceRepository>().networkServices[network.active]!;
    final tacAcceptance = context.read<TermsAndConditionAcceptance>();
    _updateValidTac(services.walletProxy, tacAcceptance);
  }
```

File: lib/screens/home/screen.dart [hunk 2]
```diff
@@ -19,25 +19,37 @@ class _HomeScreenState extends State<HomeScreen> {
...
+
   @override
   Widget build(BuildContext context) {
+    final tacAcceptance = context.watch<TermsAndConditionAcceptance>();
```
Comment (schwartz-concordium) [inline lib/screens/home/screen.dart:40]:
```suggestion
```

File: lib/screens/home/screen.dart [hunk 3]
```diff
@@ -19,25 +19,37 @@ class _HomeScreenState extends State<HomeScreen> {
...
     return Scaffold(
       body: Container(
         padding: const EdgeInsets.fromLTRB(16, 64, 16, 16),
-        child: Builder(
-          builder: (context) {
-            final tacState = context.watch<TermsAndConditionAcceptance>();
+        child: BlocConsumer<TermsAndConditionAcceptance, TermsAndConditionsAcceptanceState>(
+          bloc: tacAcceptance,
```
Comment (schwartz-concordium) [inline lib/screens/home/screen.dart:45]:
```suggestion
```

We can omit the `bloc` since `BlocConsumer` will do the lookup.

File: lib/screens/home/screen.dart (hunk 1 above)
Comment (schwartz-concordium) [reply lib/screens/home/screen.dart:36]: See comment about avoiding `bloc` parameter.

File: lib/screens/home/screen.dart (hunk 2 above)
Comment (schwartz-concordium) [reply lib/screens/home/screen.dart:40]: See comment about avoiding `bloc` parameter.

**Review by schwartz-concordium — CHANGES_REQUESTED** [review]

File: lib/screens/home/screen.dart (hunk 1 above)
Comment (bisgardo) [reply lib/screens/home/screen.dart]: done

File: lib/screens/home/screen.dart (hunk 2 above)
Comment (bisgardo) [reply lib/screens/home/screen.dart]: done

File: lib/screens/home/screen.dart (hunk 3 above)
Comment (bisgardo) [reply lib/screens/home/screen.dart]: done

File: lib/screens/terms_and_conditions/screen.dart [hunk 4]
```diff
@@ -121,11 +121,11 @@ class _TermsAndConditionsScreenState extends State<TermsAndConditionsScreen> {
 
   Function()? _onAcceptButtonPressed(BuildContext context) {
     if (isAccepted) {
-      return () async {
+      return () {
         final tac = AcceptedTermsAndConditions(
           version: widget.validTermsAndConditions.version,
         );
-        await context.read<TermsAndConditionAcceptance>().userAccepted(tac);
+        context.read<TermsAndConditionAcceptance>().userAccepted(tac);
```
Comment (bisgardo) [inline lib/screens/terms_and_conditions/screen.dart:128]: This reverts https://github.com/Concordium/concordium-wallet/pull/12#discussion_r1377305866 because the persist call is being handled in an `onChange` observer method which is called after setting the state.

But I think this is actually the right approach as we don't want async actions to block the UI. Rather, the async completion should push its status somewhere which might then trigger a UI update.

**Review by schwartz-concordium — APPROVED** [review]

File: lib/screens/home/screen.dart [hunk 5]
```diff
@@ -19,25 +19,36 @@ class _HomeScreenState extends State<HomeScreen> {
...
+    //       (now and on other appropriate triggers like activating the app).
+    _refresh(context);
   }
 
   static Future<void> _updateValidTac(WalletProxyService walletProxy, TermsAndConditionAcceptance tacAcceptance) async {
     final tac = await walletProxy.fetchTermsAndConditions();
     tacAcceptance.validVersionUpdated(ValidTermsAndConditions.refreshedNow(termsAndConditions: tac));
   }
 
+  void _refresh(BuildContext context) {
```
[the remark this answers is not in the document]
Comment (bisgardo) [reply lib/screens/home/screen.dart:32]: This is just a helper function called from a bloc listener which expects it to take care of the entire refresh. I think returning the future kind of indicates that the caller is supposed to do something with it. And that the appropriate thing to if we want the UI to react on the future is to store it as widget state.
